velox/dwio/parquet/reader/ParquetReader.cpp

```diff
@@ -644,6 +644,15 @@ std::unique_ptr<ParquetTypeWithId> ReaderBase::getParquetColumnInfo(
   } else { // leaf node
     name = columnNames.at(curSchemaIdx);
     const auto veloxType = convertType(schemaElement, requestedType);
+    // Check if the requestedType matches the veloxType created from the schema
+    // if provided.
+    if (requestedType) {
+      VELOX_CHECK(
+          veloxType->equivalent(*requestedType),
```

Comment:
This does not need to be exact match, some schema evolution should be supported (e.g. INTEGER->BIGINT).  You should add individual checks inside `convertType` for each converted type.

Reply:
This makes sense! We should confirm what schema evolution is currently supported.

Reply:
At file level we support these:
* Struct field rename
* Additional field at end of struct
* Type widening: all integer types can be widen; REAL can be widen to DOUBLE; there is no conversion between floating point types and integer types

Check out `TableEvolutionFuzzer` to see some examples (ideally we want enable it for Parquet as well): https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/TableEvolutionFuzzerTest.cpp#L140

Reply:
Struct field rename PR is here https://github.com/facebookincubator/velox/pull/5962
We need an overall design for schema evolution.

In this PR, we should at least throw a reasonable unsupported error instead of SEGV.

> You should add individual checks inside convertType() for each converted type.

Let's do this as a starting point. We can error out for unsupported schema evolution.